### PR TITLE
Write Anker metrics to prometheus

### LINKED ISSUE #450: [Anchor] Record statistics about deposits and withdrawals
In Solido we keep a histogram of deposits. This is quite useful, we should do the same for Anker. We may need to generalize the histogram over the token type then.

### PR DESCRIPTION
**Pull request by enriquefynn**
Write Anker metrics to prometheus so we can observe deposits/withdrawals

Closes #450

### CODE REVIEW

File: cli/common/src/prometheus.rs [hunk 1]
```diff
@@ -85,6 +91,18 @@ impl<'a> Metric<'a> {
         Metric::new(MetricValue::Nano(amount.0))
     }
 
+    /// Construct a metric that measures an amount of UST.
+    pub fn new_ust_sol(amount: MicroUst) -> Metric<'a> {
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:95]:
```suggestion
    pub fn new_ust(amount: MicroUst) -> Metric<'a> {
```
?

File: cli/common/src/prometheus.rs [hunk 2]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+    metrics: &anker::metrics::Metrics,
+    at: SystemTime,
+    out: &mut W,
+) -> io::Result<()> {
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_swapped_rewards_st_sol_total",
+            help: "Total amount of stSOL rewards swapped by our Anker instance.",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:339]:
```suggestion
            type_: "gauge",
```

This is data that comes from the chain, not from inside the daemon itself, right? In that case it should be a gauge, not a counter. The value might go down for example when the daemon connects to a loadbalanced RPC node, and one of the RPC nodes behind the loadbalancer is a few slots behind. Then we can get this flip-flopping behavior, and Prometheus should not treat that as a counter reset.

File: cli/common/src/prometheus.rs [hunk 3]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+            metrics: vec![Metric::new_st_sol(metrics.swapped_rewards_st_sol_total).at(at)],
+        },
+    )?;
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_swapped_rewards_ust_total",
+            help: "Total amount of UST rewards swapped by our Anker instance.",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:348]:
```suggestion
            type_: "gauge",
```
Also here.

File: cli/common/src/prometheus.rs [hunk 4]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+        },
+    )?;
+
+    // Deposit metrics
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_deposit_st_sol_total",
+            help: "Total amount stSOL deposited into Anker",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:359]:
```suggestion
            type_: "gauge",
```

File: cli/common/src/prometheus.rs [hunk 5]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+            metrics: vec![Metric::new_st_sol(metrics.deposit_metric.st_sol_total).at(at)],
+        },
+    )?;
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_deposit_b_sol_total",
+            help: "Total amount bSOL minted in response to a deposit into Anker",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:368]:
```suggestion
            type_: "gauge",
```

File: cli/common/src/prometheus.rs [hunk 6]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+            metrics: vec![Metric::new_b_sol(metrics.deposit_metric.b_sol_total).at(at)],
+        },
+    )?;
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_deposit_count_total",
+            help: "Total number of deposits made by users on Anker.",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:377]:
```suggestion
            type_: "gauge",
```

File: cli/common/src/prometheus.rs [hunk 7]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+        },
+    )?;
+
+    // Withdraw metrics
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_withdraw_st_sol_total",
+            help: "Total amount of stSOL withdrawn in response from burning bSOL from Anker",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:388]:
```suggestion
            type_: "gauge",
```

File: cli/common/src/prometheus.rs [hunk 8]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+            metrics: vec![Metric::new_st_sol(metrics.withdraw_metric.st_sol_total).at(at)],
+        },
+    )?;
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_withdraw_b_sol_total",
+            help: "Total amount of bSOL burned in response from withdrawing stSOL from Anker",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:397]:
```suggestion
            type_: "gauge",
```

File: cli/common/src/prometheus.rs [hunk 9]
```diff
@@ -307,6 +326,91 @@ pub fn write_solido_metrics_as_prometheus<W: io::Write>(
...
+            metrics: vec![Metric::new_b_sol(metrics.withdraw_metric.b_sol_total).at(at)],
+        },
+    )?;
+    write_metric(
+        out,
+        &MetricFamily {
+            name: "anker_withdraw_count_total",
+            help: "Total number of withdrawals made by users on Anker.",
+            type_: "counter",
```
Comment (ruuda) [inline cli/common/src/prometheus.rs:406]:
```suggestion
            type_: "gauge",
```

**Review by ruuda — COMMENTED** [review]
I have a few small comments, LGTM aside from those!

File: cli/common/src/prometheus.rs (hunk 1 above)
Comment (enriquefynn) [reply cli/common/src/prometheus.rs:95]: Copy-paste bug

**Pull request merged**